controllers/metal3labelsync_controller.go

```diff
@@ -0,0 +1,387 @@
+	labelSyncSet := make(map[string]string)
+	for labelKey, labelVal := range labels {
+		p, n := k8strings.SplitQualifiedName(labelKey)
+		if p == "" || n == "" {
+			continue
+		}
+		if _, ok := prefixSet[p]; !ok {
+			continue
+		}
+		labelSyncSet[labelKey] = labelVal
+	}
+	return labelSyncSet
+}
+
+func synchronizeLabelSyncSetsOnNode(hostLabelSyncSet, nodeLabelSyncSet map[string]string, node *corev1.Node) {
+	if node.Labels == nil {
+		node.Labels = map[string]string{}
+	}
+	for labelKey, labelVal := range nodeLabelSyncSet {
+		val, ok := hostLabelSyncSet[labelKey]
+		if !ok || val != labelVal {
+			delete(node.Labels, labelKey)
+		}
+	}
+	for labelKey, labelVal := range hostLabelSyncSet {
+		val, ok := nodeLabelSyncSet[labelKey]
+		if !ok || val != labelVal {
+			node.Labels[labelKey] = labelVal
+		}
+	}
+}
+
+// SetupWithManager will add watches for this controller
+func (r *Metal3LabelSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&bmh.BareMetalHost{}).
+		Watches(
+			&source.Kind{Type: &capm3.Metal3Cluster{}},
+			&handler.EnqueueRequestsFromMapFunc{
+				ToRequests: handler.ToRequestsFunc(r.Metal3ClusterToBareMetalHosts),
+			},
+		).
+		Complete(r)
+}
+
+// Metal3ClusterToBareMetalHosts is a handler.ToRequestsFunc to be used to enqeue
+// requests for reconciliation of BareMetalHosts' label updates.
+func (r *Metal3LabelSyncReconciler) Metal3ClusterToBareMetalHosts(o handler.MapObject) []ctrl.Request {
+	result := []ctrl.Request{}
+	c, ok := o.Object.(*capm3.Metal3Cluster)
+	if !ok {
+		r.Log.Error(errors.Errorf("expected a Metal3Cluster but got a %T", o.Object),
+			"failed to get BareMetalHost for Metal3Cluster",
+		)
+		return nil
+	}
+	log := r.Log.WithValues("Metal3Cluster", c.Name, "Namespace", c.Namespace)
+	cluster, err := util.GetOwnerCluster(context.TODO(), r.Client, c.ObjectMeta)
+	switch {
+	case apierrors.IsNotFound(err) || cluster == nil:
+		return nil
+	case err != nil:
+		log.Error(err, "failed to get owning cluster")
+		return nil
+	}
+
+	labels := map[string]string{capi.ClusterLabelName: cluster.Name}
+	capiMachineList := &capi.MachineList{}
+	if err := r.Client.List(context.TODO(), capiMachineList, client.InNamespace(c.Namespace), client.MatchingLabels(labels)); err != nil {
+		log.Error(err, "failed to list Machines")
+		return nil
+	}
+	for _, m := range capiMachineList.Items {
+		if m.Spec.InfrastructureRef.Name == "" {
+			continue
+		}
+		name := client.ObjectKey{Namespace: m.Namespace, Name: m.Spec.InfrastructureRef.Name}
+		if m.Spec.InfrastructureRef.Namespace != "" {
+			name = client.ObjectKey{Namespace: m.Spec.InfrastructureRef.Namespace, Name: m.Spec.InfrastructureRef.Name}
+		}
+		capm3Machine := &capm3.Metal3Machine{}
+		if err := r.Client.Get(context.TODO(), name, capm3Machine); err != nil {
+			log.Error(err, "failed to get Metal3Machine")
+			return nil
```

Comment:
I'm wondering if we should rather use "continue" here to not block the reconciliation on one machine. For example a user has scaled out a machine deployment but does not have enough BMH. If that machine is first in the list, it will always return on line 311. Using continue, would allow us to get past that.

Reply:
@maelk That's a good point. I had thought about this too and took the more conservative route. In addition to line 311, should we also use continue for 316 and 321?

Reply:
yes, I think for all exit conditions, we should instead use "continue"

Reply:
Done.

Reply:
what about line 306. There could be situations (though hopefully not frequent) where a machine was created without metal3machine (yet ?). We should probably use continue there too, no ?

Reply:
@maelk The if statement above "if m.Spec.InfrastructureRef.Name == nil", should catch the case where Machine exists but Metal3Machine does not. I suppose there is a rare corner case where InfrastructureRef is updated but Metal3Machine is not yet created. I will make the change.